bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java

```diff
@@ -0,0 +1,173 @@
+		initializeComboContents();
+		fillComments();
+	}
+	
+	private void initializeComboContents() {
+		cmbEditPriority.setItems(model.getPriorties());
+		cmbEditType.setItems(model.getTypes());
+		cmbEditStatus.setItems(model.getStatuses());
+		cmbEditAssignedTo.setItems(model.findReporters());
+		cmbEditVersion.setItems(model.findProjectVersions(model.getReportForEdit().getProject()));
+	}
+	
+	private void fillComments() {
+		VerticalLayout layout = (VerticalLayout) pnlCommentsThread.getContent();
+		layout.removeAllComponents();
+		List<Comment> comments = model.getComments();
+		for (Comment comment : comments) {
+			layout.addComponent(new ThreadItemView(new ThreadItemModel(comment)));
+		}
+	}
+	
+	
+	private void saveReport() {
+		model.save();
+		Notification.show("Saved successfully", Type.TRAY_NOTIFICATION);
+	}
+	
+	private void revertChanges() {
+		model.resetReportForEdit();
+		binder.setBean(model.getReportForEdit());
+	}
+
+	private void saveComment() {
+		if (!txtComment.getOptionalValue().isPresent())
+			return;
+		model.saveComment(txtComment.getValue(), BaseModel.loginUser);
+		fillComments();
+		txtComment.clear();
+	}
+
+
+	@Override
+	public void uploadStarted(StartedEvent event) {
+		pnlAttachments.setVisible(true);
+		HorizontalLayout layout = (HorizontalLayout) pnlAttachments.getContent();
+		ProgressBar progress = new ProgressBar();
+		progress.setCaption(event.getFilename());
+		model.getAttachmentUIElements().put(event.getFilename(), progress);
+		layout.addComponent(progress);
+	}
+
+
+	@Override
+	public void uploadSucceeded(SucceededEvent event) {
+		ProgressBar progress = (ProgressBar) model.getAttachmentUIElements().get(event.getFilename());
+		HorizontalLayout layout = (HorizontalLayout) pnlAttachments.getContent();
+		layout.removeComponent(progress);
+		
+		FileResource file = new FileResource(new File(ReportsModel.FILEUPLOAD_PATH + event.getFilename()));
+		Link link = new Link(event.getFilename(), file);
+		link.setStyleName("tiny");
+		layout.addComponent(link);
+		model.getAttachmentUIElements().put(event.getFilename(), link);
+		try {
+			model.saveAttachment(event.getFilename(), event.getMIMEType(), file.getStream());
+		} catch (IOException e) {
+			Notification.show("Error occurred while uploading the file", e.getMessage(), Type.ERROR_MESSAGE);
+		}
+		//Refresh comments after new comment;
+		fillComments();
+	}
+	
+
+	@Override
+	public OutputStream receiveUpload(String filename, String mimeType) {
+		FileOutputStream fos = null;
+		try {
+			fos = new FileOutputStream(ReportsModel.FILEUPLOAD_PATH + filename);
+		} catch (FileNotFoundException e) {
+			e.printStackTrace();
+		} finally {
+			return fos;
+		}
+	}
+
+
+	@Override
+	public void updateProgress(long readBytes, long contentLength) {
+		ProgressBar bar = null;
+		for (Object obj : model.getAttachmentUIElements().values()) {
+			if (obj instanceof ProgressBar)
+				bar = (ProgressBar)obj;
+		}
+		if (bar != null)
```

Comment:
Is always only the last progress bar being updated here?

Reply:
No changes here yet. Seems that it continues to update only the last progress bar. What if we have two concurrent uploads?